### PR TITLE
Fix/stripping of prompts

### LINKED ISSUE #194: [BUG] Deal with spaces when dealing with classification datasets
**Describe the bug**
When benchmarking on text classification datasets then we strip spaces at the end of the prompt, resulting in things like:

`<document>\nSentiment:`

This means that some models will have a high probability of a whitespace token " ". We have the opposite problem if we include a space, resulting in the following prompt:

`<document>\nSentiment: `

In this case, some models have the tokens ` positive`, ` neutral` and ` negative` (note the whitespace).

What we can do is tokenize the labels in advance, with and without a preceeding space, and add the space that we would need to make the evaluation more fair for the model.

**To Reproduce**
Run `scandeval -m ai-forever/mGPT -d swerec` and check the file `.scandeval_cache/model_cache/ai-forever--mGPT/swerec-model-outputs.json` (search for "completion").

**Context (please complete the following information):**
 - OS [e.g., MacOS, Ubuntu]: Ubuntu
 - Python version [e.g., 3.10.6]: 3.10.12
 - Device [e.g., CUDA, CPU, MPS]: CUDA
 - Output of `pip list` or `poetry show`: 
 ```
(.venv) ubuntu@man-in-pink:~$ pip list
Package                   Version
------------------------- ------------
absl-py                   2.0.0
accelerate                0.25.0
aiohttp                   3.9.1
aioprometheus             23.12.0
aiosignal                 1.3.1
annotated-types           0.6.0
anyio                     4.2.0
async-timeout             4.0.3
attrs                     23.2.0
bert-score                0.3.13
bitsandbytes              0.41.3.post2
certifi                   2023.11.17
charset-normalizer        3.3.2
chex                      0.1.85
click                     8.1.7
contourpy                 1.2.0
cycler                    0.12.1
datasets                  2.16.1
demjson3                  3.0.6
dill                      0.3.7
einops                    0.7.0
etils                     1.6.0
evaluate                  0.4.1
exceptiongroup            1.2.0
fastapi                   0.109.0
filelock                  3.13.1
flash-attn                2.4.2
flax                      0.7.5
fonttools                 4.47.0
frozenlist                1.4.1
fsspec                    2023.10.0
h11                       0.14.0
httptools                 0.6.1
huggingface-hub           0.20.1
idna                      3.6
importlib-resources       6.1.1
interegular               0.3.3
jax                       0.4.23
jaxlib                    0.4.23
Jinja2                    3.1.2
joblib                    1.3.2
jsonschema                4.21.1
jsonschema-specifications 2023.12.1
kiwisolver                1.4.5
Levenshtein               0.23.0
lm-format-enforcer        0.8.2
markdown-it-py            3.0.0
MarkupSafe                2.1.3
matplotlib                3.8.2
mdurl                     0.1.2
ml-dtypes                 0.3.2
mpmath                    1.3.0
msgpack                   1.0.7
multidict                 6.0.4
multiprocess              0.70.15
nest-asyncio              1.5.8
networkx                  3.2.1
ninja                     1.11.1.1
nltk                      3.8.1
numpy                     1.26.3
nvidia-cublas-cu12        12.1.3.1
nvidia-cuda-cupti-cu12    12.1.105
nvidia-cuda-nvrtc-cu12    12.1.105
nvidia-cuda-runtime-cu12  12.1.105
nvidia-cudnn-cu12         8.9.2.26
nvidia-cufft-cu12         11.0.2.54
nvidia-curand-cu12        10.3.2.106
nvidia-cusolver-cu12      11.4.5.107
nvidia-cusparse-cu12      12.1.0.106
nvidia-nccl-cu12          2.18.1
nvidia-nvjitlink-cu12     12.3.101
nvidia-nvtx-cu12          12.1.105
openai                    0.28.1
opt-einsum                3.3.0
optax                     0.1.7
orbax-checkpoint          0.4.8
orjson                    3.9.12
packaging                 23.2
pandas                    1.5.3
pillow                    10.2.0
pip                       23.3.2
protobuf                  3.20.3
psutil                    5.9.7
pyarrow                   14.0.2
pyarrow-hotfix            0.6
pydantic                  1.10.13
pydantic_core             2.14.6
Pygments                  2.17.2
pyinfer                   0.0.3
pyparsing                 3.1.1
python-dateutil           2.8.2
python-dotenv             0.21.1
pytz                      2023.3.post1
PyYAML                    6.0.1
quantile-python           1.1
rapidfuzz                 3.6.1
ray                       2.9.1
referencing               0.32.1
regex                     2023.12.25
requests                  2.31.0
responses                 0.18.0
rich                      13.7.0
rouge_score               0.1.2
rpds-py                   0.17.1
sacremoses                0.1.1
safetensors               0.4.1
scandeval                 9.2.0
scikit-learn              1.3.2
scipy                     1.11.4
sentencepiece             0.1.99
seqeval                   1.2.2
setuptools                59.6.0
six                       1.16.0
sniffio                   1.3.0
starlette                 0.35.1
sympy                     1.12
tabulate                  0.9.0
tensorstore               0.1.52
termcolor                 2.4.0
threadpoolctl             3.2.0
tiktoken                  0.5.2
tokenizers                0.15.0
toolz                     0.12.0
torch                     2.1.2
tqdm                      4.66.1
transformers              4.36.2
triton                    2.1.0
typing_extensions         4.9.0
urllib3                   2.1.0
uvicorn                   0.27.0
uvloop                    0.19.0
vllm                      0.2.7
watchfiles                0.21.0
websockets                12.0
wheel                     0.42.0
xformers                  0.0.23.post1
xxhash                    3.4.1
yarl                      1.9.4
zipp                      3.17.0
```

### PR DESCRIPTION
**Pull request by saattrupdan**
## Summary
Main thing in this PR is dealing with when we're stripping the prompts. A prompt can look like the following:

`Tweet: <some tweet>\nSentiment:`

Note that there's no space after the last colon. When we're asking the model for the next token then it might output the token ` positive` (note the prefix space), which is great, but it might also simply output a whitespace token ` `, with `positive` following that. This is an issue when we're looking at the probability distribution of the next token, since we want the distribution of the next "meaningful" token.

To counter this, we tokenise `positive` and ` positive` separately, and check if the tokens of `positive` are included in the tokens of ` positive`. If this is the case, then we _should not_ strip the prompts, since the initial whitespace would be recorded as a separate token if we strip.

As an example of this, the tokens of `positive` could, e.g., be [`pos`, `itive`], and the tokens of ` positive` could be [` `, `pos`, `itive`]. In this case we _do not_ want to strip the prompt, since we would just get the whitespace ` ` token as the first generated token in that case. As another example, the tokens of `positive` could be [`po`, `sitive`] and the tokens of ` positive` be [` positive`], in which case we _do_ want to strip the prompt, since we care about the probability of ` positive`.

### Changed
- Number of generated tokens for sequence classification tasks has been changed back to 3 (from 1). This makes no difference to open source models, as we only use the logprobs from the first token anyway, but it *does* make a difference to closed source models where the logprobs are not available (like OpenAI's chat models), as we're instead calculating word edit distance to the labels.

### Fixed
- We now only strip the prompts if the model's tokenizer includes a prefix space when tokenizing the labels.

This closes #194.

### CODE REVIEW

File: src/scandeval/benchmark_dataset.py [hunk 1]
```diff
@@ -468,6 +483,23 @@ def _load_prepared_data(
...
+                        # token
+                        labels_to_be_generated = list(
+                            self.dataset_config.prompt_label_mapping.values()
+                        )
+                        strip_prompts = should_prompts_be_stripped(
+                            labels_to_be_generated=labels_to_be_generated,
+                            tokenizer=tokenizer,
+                        )
+                        if strip_prompts:
+                            test = test.map(
```
Comment (KennethEnevoldsen) [inline src/scandeval/benchmark_dataset.py:497]: I am unsure what test is here - I would rename it "prompt_dataset" (assuming that is what it is).

**Review by KennethEnevoldsen — APPROVED** [review]
Good catch! (how did you notice?)

only a minor thing. Otherwise looks good.

File: src/scandeval/benchmark_dataset.py (hunk 1 above)
Comment (saattrupdan) [reply src/scandeval/benchmark_dataset.py:497]: It's the test split :) 

Prompt dataset is misleading, since this both for decoders and encoders

**Comment by saattrupdan** [comment]
> Good catch! (how did you notice?)
> 
> 
> 
> only a minor thing. Otherwise looks good.

Just noticed while browsing the outputs of the models during evaluation, really 🙃

File: src/scandeval/benchmark_dataset.py (hunk 1 above)
Comment (KennethEnevoldsen) [reply src/scandeval/benchmark_dataset.py:497]: Ah right - then that is my misunderstanding keep it as is.

**Pull request merged**